src/configuration/File.h

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (C) 1996-2021 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#ifndef _SQUID__SRC_CONFIGURATION_FILE_H
+#define _SQUID__SRC_CONFIGURATION_FILE_H
+
+#include "base/Here.h"
+#include "configuration/forward.h"
+#include "sbuf/SBuf.h"
+
+#include <cstdio>
+#include <string>
+
+namespace Configuration
+{
+
+/**
+ * Class used to store required information from a
+ * configuration file.
+ */
+class File
```

Comment:
To facilitate moving this forward:

> > Alex: Which of the following four configuration file categories do you want to use this class for?
> 
>     1. the primary configuration file (squid.conf or equivalent)

A `Configuration::File` presents line-wise strings to `parseOneConfigFile()`

>     2. configuration files included using the `include` directive

`parseManyConfigFiles()` iterates across the included filenames, passing them to (1) for access via `Configuration::File`

>     3. directive parameter files named by the "quoted-filename" construct in a `acl` directive

`ConfigParser::CfgFile` opens the file with `Configuration::File` and translates the line-wise strings into elements.

>     4. directive parameter files named by the `parameters(filename)` construct

`ConfigParser::CfgFile` opens the file with `Configuration::File` and translates the line-wise strings into elements.

>     5. Added by Amos: command line specified snippets of configuration
> 
>     6. Added by Amos: cache manager POST/PUT request payloads

Concept for the above is to attach a `Configuration::File` to network socket, dump any pre-read data into its buffer and have it interpret the bytes for presentation to the config parser.

> 
> OK, it is possible to create a class that would just store raw configuration bytes, of course. Such a class is unlikely to be much more than a (name,buffer) tuple, essentially, but that does not mean it cannot be useful.
> 

squid.conf parser is (still) a multi-layered beast. This is only intended to replace the redundant fundamental file I/O layer(s) of it with a single unified object interpreting config files at their lowest common abstraction - which is lines. From that it can filter out the human-only content from the config files. comments, empty lines, line-wrap, etc.

For now we still have multiple parsers using this `Configuration::File` object in the ways I outline above to interpret the syntax of those lines differently. Fixing / unifying those is a future PR project.

Reply:
Please keep in mind that this is just one step of a refactoring project. Starting with a complete up-front design of the entire feature has proven time and again to be a fatal flaw working with Squid code. Not least because all these design requirements and/or guarantees you insist on baking into the code at the start (as above). IME we will hit some tiny detail later in the process which we overlooked here and you will veto the final stages of refactor because it violates the guarantees insisted on here.

I would rather like to leave these things flexible for later refactor updates. The current implementation only handles files as data sources, thus the class naming. If/when that changes in future the class can easily be renamed.
 n summary - This PR code is "good enough" and correct for current state of Squid.